parmed/modeller/residue.py

```diff
@@ -329,6 +377,100 @@ def fix_charges(self, to=None, precision=4):
 
         return self
 
+    def apply_patch(self, patch, precision=4):
+        """
+        Apply the specified PatchTemplate to the ResidueTemplate.
+
+        This only handles patches that affect a single residue.
+
+        An exception is thrown if patch is incompatible because
+        * The patch specifies that an atom is to be deleted that doesn't exist in the residue
+        * A bond specified as being added in the patch does not have both atom names present after adding/deleting atoms from the patch
+        * The new net charge is not integral to the specified precision
+
+        Parameters
+        ----------
+
+        patch : PatchTemplate
+            The patch to apply to this residue
+
+        precision : int, optional
+            Each valid patch should be produce a net charge that is integral to
+            this many decimal places.
+            Default is 4
+
+        Returns
+        -------
+
+        residue : ResidueTemplate
+            A new ResidueTemplate corresponding to the patched residue is returned.
+            The original remains unmodified.
+
+        """
+        # Create a copy
+        # TODO: Once ResidueTemplate.from_residue() actually copies all info, use that instead?
+        residue = _copy.copy(self)
+        # Delete atoms
+        for atom_name in patch.delete_atoms:
+            residue.delete_atom(atom_name)
+        # Add or replace atoms
+        for atom in patch.atoms:
+            if atom.name in residue:
+                # Overwrite type and charge
+                residue[atom.name].type = atom.type
+                residue[atom.name].charge = atom.charge
+            else:
+                residue.add_atom(Atom(name=atom.name, type=atom.type, charge=atom.charge))
+        # Add bonds
+        for (atom1_name, atom2_name, order) in patch.add_bonds:
+            try:
+                # Remove dangling bonds
+                for name in [atom1_name, atom2_name]:
+                    if name == residue.head:
+                        residue.head = None
+                    if name == residue.tail:
+                        residue.tail = None
+                # Add bond
+                residue.add_bond(atom1_name, atom2_name, order)
+            except:
+                raise Exception('Bond %s-%s could not be added to patched residue: atoms are %s' % (atom1_name, atom2_name, list(residue._map.keys())))
```

Comment:
Create a new exception type for this in `parmed.exceptions` and raise that.  Raising `Exception` requires you to catch it with an `except` that necessarily catches everything, making it too likely that you'll wind up masking a bug.

Reply:
Adding `IncompatiblePatchError`